File: scripts/validate_stations.py
import re
import json
import urllib.request
import urllib.error
import sys
from concurrent.futures import ThreadPoolExecutor, as_completed

# 1. Parse tflStations.ts to get all station ids
with open('data/tflStations.ts', 'r') as f:
    tfl_stations_content = f.read()

# Find all { id: '...', name: '...' } patterns in TFL_STATIONS
station_ids = re.findall(r"id:\s*'([^']+)'", tfl_stations_content)
station_names = re.findall(r"name:\s*[\"']([^\"']+)[\"']", tfl_stations_content)

stations = list(zip(station_ids, station_names))
print(f"Loaded {len(stations)} curated legacy station slugs from tflStations.ts")

# 2. Parse resolveTflStopId.ts to build maps
with open('utils/resolveTflStopId.ts', 'r') as f:
    resolver_content = f.read()

# Load fullStationsData JSON
with open('data/tflStationsFull.json', 'r') as f:
    full_stations_data = json.load(f)

# Build SLUG_TO_NAPTAN simulation
def to_slug(name):
    slug = name.lower()
    slug = re.sub(r'[^a-z0-9]+', '-', slug)
    slug = re.sub(r'^-+|-+$', '', slug)
    return slug

SLUG_TO_NAPTAN = {}
for entry in full_stations_data:
    entry_id = entry.get('id', '')
    if entry_id.startswith('940GZZ') or entry_id.startswith('910G'):
        slug = to_slug(entry.get('name', ''))
        existing = SLUG_TO_NAPTAN.get(slug)
        if not existing:
            SLUG_TO_NAPTAN[slug] = entry_id
        else:
            if existing.startswith('910G') and entry_id.startswith('940GZZ'):
                SLUG_TO_NAPTAN[slug] = entry_id

# Extract SLUG_TO_HUB map
slug_to_hub_block = re.search(r'const SLUG_TO_HUB: Record<string, string> = \{(.*?)\};', resolver_content, re.DOTALL)
SLUG_TO_HUB = {}
if slug_to_hub_block:
    for line in slug_to_hub_block.group(1).split('\n'):
        line = line.strip()
        if line and ':' in line:
            parts = line.split(':')
            k = parts[0].strip().replace("'", "").replace('"', '')
            v = parts[1].strip().replace("'", "").replace('"', '').replace(',', '')
            SLUG_TO_HUB[k] = v

# Extract EXPLICIT_MAP map
explicit_map_block = re.search(r'const EXPLICIT_MAP: Record<string, string\[\]> = \{(.*?)\};', resolver_content, re.DOTALL)
EXPLICIT_MAP = {}
if explicit_map_block:
    for line in explicit_map_block.group(1).split('\n'):
        line = line.strip()
        if line and ':' in line:
            parts = line.split(':')
            k = parts[0].strip().replace("'", "").replace('"', '')
            arr_str = parts[1].strip()
            arr_matches = re.findall(r"['\"]([^'\"]+)['\"]", arr_str)
            EXPLICIT_MAP[k] = arr_matches

def resolve_tfl_stop_ids(station_id):
    if station_id.startswith('HUB') or station_id.startswith('940GZZ') or station_id.startswith('910G'):
        return [station_id]
    
    slug = to_slug(station_id)
    if slug in EXPLICIT_MAP:
        return EXPLICIT_MAP[slug]
    if station_id in EXPLICIT_MAP:
        return EXPLICIT_MAP[station_id]
        
    if slug in SLUG_TO_NAPTAN:
        return [SLUG_TO_NAPTAN[slug]]
        
    return [station_id]

# 3. Verify all stations & hit live Vercel backend in parallel
backend_base = "https://my-commute-brain.vercel.app/api/stations"
failed_resolution = []
failed_live = []

print("\n--- RESOLUTION & LIVE BACKEND API CHECK ---")

def verify_single_id(sid, sname, rid):
    url = f"{backend_base}/{rid}"
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req, timeout=10) as response:
            data = json.loads(response.read().decode())
            if 'departures' not in data:
                raise ValueError("JSON missing 'departures' property")
            dept_count = len(data.get('departures', []))
            return (True, sid, sname, rid, dept_count, None)
    except Exception as e:
        return (False, sid, sname, rid, 0, f"{url} -> {e}")

tasks = []
for sid, sname in stations:
    resolved = resolve_tfl_stop_ids(sid)
    is_resolved = all(r.startswith('940GZZ') or r.startswith('910G') or r.startswith('HUB') for r in resolved)
    
    if not is_resolved:
        failed_resolution.append((sid, sname, resolved))
        print(f"❌ '{sname}' (id: '{sid}') failed resolution: resolved to {resolved}")
        continue

    for rid in resolved:
        tasks.append((sid, sname, rid))

# Execute live calls in parallel (20 threads)
with ThreadPoolExecutor(max_workers=20) as executor:
    futures = {executor.submit(verify_single_id, t[0], t[1], t[2]): t for t in tasks}
    for future in as_completed(futures):
        success, sid, sname, rid, dept_count, err_msg = future.result()
        if success:
            print(f"✅ '{sname}' -> '{rid}' works ({dept_count} live departures)")
        else:
            failed_live.append((sid, sname, rid, err_msg))
            print(f"❌ '{sname}' -> '{rid}' failed live call: {err_msg}")

print("\n--- FINAL AUDIT SUMMARY ---")
if not failed_resolution and not failed_live:
    print("✅ SUCCESS: 100% of stations resolve and fetch correctly from live backend!")
    sys.exit(0)
else:
    if failed_resolution:
        print(f"⚠️ {len(failed_resolution)} stations failed resolution mappings.")
    if failed_live:
        print(f"⚠️ {len(failed_live)} stations failed live backend API validation.")
    sys.exit(1)

File: scripts/validate_stations.js
const fs = require('fs');
const path = require('path');
const { execFile } = require('child_process');
const { promisify } = require('util');
const execFilePromise = promisify(execFile);

console.log('--- UPGRADED INTEGRATION TEST: DYNAMIC EVALUATION & EXPECTED-LINES ---');

// 1. Evaluate tflStations.ts dynamically
const tflStationsContent = fs.readFileSync(path.join(__dirname, '../data/tflStations.ts'), 'utf8');
const cleanTflStationsJs = tflStationsContent
  .replace(/import\s+[^;]+;/g, '') // Strip imports
  .replace(/export\s+interface\s+\w+\s*\{[^}]*\}/g, '') // Strip interfaces
  .replace(/:\s*TfLStation(\[\])?/g, '') // Strip types
  .replace(/:\s*string(\[\])?/g, '') // Strip string parameter/return types
  .replace(/as\s+[^)]+/g, '') // Strip type assertions inside parentheses
  .replace(/export\s+/g, ''); // Strip exports

const fullStationsData = JSON.parse(fs.readFileSync(path.join(__dirname, '../data/tflStationsFull.json'), 'utf8'));

// Evaluate in sandbox context
const tflContext = new Function('fullStationsData', cleanTflStationsJs + '; return { TFL_STATIONS };');
const { TFL_STATIONS } = tflContext(fullStationsData);

console.log(`Loaded ${TFL_STATIONS.length} stations from tflStations.ts`);

// 2. Evaluate resolveTflStopId.ts dynamically
const resolverContent = fs.readFileSync(path.join(__dirname, '../utils/resolveTflStopId.ts'), 'utf8');
const cleanResolverJs = resolverContent
  .replace(/import\s+[^;]+;/g, '') // Strip imports
  .replace(/export\s+/g, '') // Strip exports
  .replace(/:\s*Record<[^>]+>/g, '') // Strip Record definitions
  .replace(/:\s*string(\[\])?/g, '') // Strip string parameters/returns
  .replace(/:\s*number/g, '') // Strip number parameters/returns
  .replace(/as\s+[^)]+/g, ''); // Strip type assertions inside parentheses

const hubExpansions = JSON.parse(fs.readFileSync(path.join(__dirname, '../data/hubExpansions.json'), 'utf8'));

const resolverContext = new Function('hubExpansions', 'fullStationsData', cleanResolverJs + '; return { resolveTflStopIds };');
const { resolveTflStopIds } = resolverContext(hubExpansions, fullStationsData);

// 3. Expected lines map to backend line_id strings
const LINE_MAP = {
  'circle': ['circle'],
  'district': ['district'],
  'hammersmith-city': ['hammersmith-city'],
  'victoria': ['victoria'],
  'northern': ['northern'],
  'piccadilly': ['piccadilly'],
  'dlr': ['dlr'],
  'elizabeth': ['elizabeth'],
  'overground': ['liberty', 'lioness', 'mildmay', 'suffragette', 'weaver', 'windrush', 'overground'],
  'bakerloo': ['bakerloo'],
  'metropolitan': ['metropolitan'],
  'waterloo-city': ['waterloo-city'],
};

// 4. Sequential execution of tests
const backendBase = 'https://my-commute-brain.vercel.app/api/stations';
const failedResolution = [];
const failedLines = [];
const failedLive = [];
const successful = [];

async function runAll() {
  const date = new Date();
  const londonTime = new Intl.DateTimeFormat('en-GB', {
    timeZone: 'Europe/London',
    hour: '2-digit',
    hour12: false
  }).format(date);
  const hour = parseInt(londonTime, 10);
  // Core London operating hours
  const isOperatingHours = hour >= 6 && hour < 23;

  for (let i = 0; i < TFL_STATIONS.length; i++) {
    const station = TFL_STATIONS[i];
    let resolved;
    try {
      resolved = resolveTflStopIds(station.id);
    } catch (e) {
      failedResolution.push({ station, error: e.message });
      console.log(`[${i + 1}/${TFL_STATIONS.length}] ❌ Resolution crashed for "${station.name}" (${station.id}): ${e.message}`);
      continue;
    }

    const isResolved = resolved.every(r => r.startsWith('940GZZ') || r.startsWith('910G') || r.startsWith('HUB'));
    
    if (!isResolved) {
      failedResolution.push({ station, resolved });
      console.log(`[${i + 1}/${TFL_STATIONS.length}] ❌ Resolution failed for "${station.name}" (${station.id}) -> resolved to ${JSON.stringify(resolved)}`);
      continue;
    }

    console.log(`[${i + 1}/${TFL_STATIONS.length}] Testing "${station.name}" -> resolved to ${JSON.stringify(resolved)}`);
    
    // Fetch all children in parallel and merge departures
    let mergedDepartures = [];
    let fetchErrors = [];
    
    await Promise.all(resolved.map(async (rid) => {
      const url = `${backendBase}/${rid}`;
      try {
        const { stdout } = await execFilePromise('/usr/bin/curl', ['-s', url]);
        const data = JSON.parse(stdout);
        if (data && Array.isArray(data.departures)) {
          mergedDepartures = mergedDepartures.concat(data.departures);
        }
      } catch (err) {
        fetchErrors.push(`${rid}: ${err.message}`);
      }
    }));

    if (fetchErrors.length > 0 && mergedDepartures.length === 0) {
      failedLive.push({ station, errors: fetchErrors });
      console.log(`   ❌ API fetch failed: ${fetchErrors.join(', ')}`);
      continue;
    }

    // Verify expected lines are present during operating hours.
    // If the board is completely empty, it's usually a rate-limit/network flake, not an expansion bug.
    // So we only assert missing lines if we successfully fetched at least one departure.
    const missingLines = [];
    const isPureRail = resolved.every(rid => rid === '910GCTMSLNK' || rid === '910GFENCHRS');
    
    if (isOperatingHours && !isPureRail) {
      if (mergedDepartures.length > 0) {
        for (const expectedLine of station.lines) {
          if (station.id === 'kensington-oly' && expectedLine === 'district') {
            continue; // District line service to Kensington Olympia is intermittent
          }
          if (expectedLine === 'waterloo-city') {
            const day = new Date().getDay();
            if (day === 0 || day === 6) {
              continue; // Waterloo & City line is closed on weekends
            }
          }
          const matchingLineIds = LINE_MAP[expectedLine] || [expectedLine];
          const hasLineArrival = mergedDepartures.some(dep => 
            dep.line_id && matchingLineIds.includes(dep.line_id.toLowerCase().trim())
          );
          if (!hasLineArrival) {
            missingLines.push(expectedLine);
          }
        }
      } else {
        console.log(`   ⚠️ Warning: Empty arrivals returned (possible network/TfL rate-limit flake)`);
      }
    }

    if (missingLines.length > 0) {
      console.log(`   ⚠️ Warning: Expected-lines check failed. Missing: ${JSON.stringify(missingLines)} (could be a planned closure/disruption)`);
      failedLines.push({ station, missingLines });
    } else {
      successful.push(station);
      console.log(`   ✅ Pass! (${mergedDepartures.length} live departures verified)`);
    }

    // 300ms sleep to avoid TfL rate limits
    await new Promise((r) => setTimeout(r, 300));
  }

  console.log('\n--- VERIFICATION RESULTS ---');
  console.log(`Total stations checked: ${TFL_STATIONS.length}`);
  console.log(`Successful: ${successful.length}`);
  console.log(`Resolution failures: ${failedResolution.length}`);
  console.log(`API fetch failures: ${failedLive.length}`);
  console.log(`Expected-lines assertion failures: ${failedLines.length}`);

  if (failedResolution.length === 0 && failedLive.length === 0 && failedLines.length === 0) {
    console.log('🎉 PASS: All configured stations resolve, fetch, and contain all expected lines!');
    process.exit(0);
  } else {
    console.log('❌ FAIL: Some stations had resolution, API, or missing-line errors.');
    process.exit(1);
  }
}

runAll();
